### DISCUSSION TITLE
When trying to retrieve the value of a pubkey using the Value Attribute the vector is coming empty

### DISCUSSION THREAD

**Discussion by shivam560** [parallaxsecond/rust-cryptoki#137] [Q&A] [answered]
```
pub fn create_key_pair() -> Result<Vec<u8>,()> {

    let (pkcs111,slit) = init_pins();
    let session = pkcs111.open_rw_session(slit).unwrap();
    session.login(cryptoki::session::UserType::User, Some("1234"));
  
    let class = ObjectClass::PUBLIC_KEY;
    let temp = vec![
        
        Attribute::Class(class),
        Attribute::Label("mykey".as_bytes().try_into().unwrap()),
      Attribute::KeyType(KeyType::RSA),
      
       
    ];
    
    let publickeyhandle =  session.find_objects(&temp).unwrap().pop().expect("No public key found in slot");
    println!("The pubkeyhandle is {}",publickeyhandle);
    let new_pub_key = session.get_attributes(publickeyhandle, &[AttributeType::CheckValue]).unwrap();
    println!("The length is {}",new_pub_key.len());
    let value = if let Attribute::Value(value) = new_pub_key[0].clone() {
        value
    } else {
        panic!("Expected value attribute.");
    };  
 
         Ok(value)
}
```

In the above code pubkeyhandle is coming as 2 
but the new_pub_key is an empty vector.

in terminal when i am checking for objects I see slot 0 is initialized with pubkey 

```
pkcs11-tool --module /usr/lib/softhsm/libsofthsm2.so --slot-index 0 --list-objects --type pubkey

Using slot with index 0 (0x1c85df1e)
Public Key Object; RSA 2048 bits
  label:      mykey
  Usage:      encrypt, verify, wrap
  Access:     local


```
But still new_pub_key is still coming as an empty vector

**Comment by shivam560** [comment]
@baloo @vkkoskie @wiktor-k @gowthamsk-arm 
I could use some help here. I am getting an empty vector when calling get_attributes() using Attribute::Value. i  generated the keys using the pkcs11-tool command in ubuntu

**Comment by baloo** [comment, marked as answer]
You are requesting the `CHECK_VALUE` attribute in the `new_pub_key` variable:
  https://docs.oasis-open.org/pkcs11/pkcs11-spec/v3.1/cs01/pkcs11-spec-v3.1-cs01.html#_Toc111203232
```
CKA_CHECK_VALUE | Byte   array | Key   checksum
```

I'm not sure this what you intend to use.
I think you're looking for `CKA_MODULUS` and `CKA_PUBLIC_EXPONENT` for an RSA key

**Reply by ionut-arm** [reply]
Indeed, this sounds right to me.

@shivam560 - you can see an example of how we get existing keys in Parsec [here](https://github.com/parallaxsecond/parsec/blob/main/src/providers/pkcs11/key_management.rs#L25). We just populate the ID attribute (which we know from when we generated the key), but if you don't know that (and for some reason `pkcs11-tool` doesn't tell you the ID........) you can just populate `Attribute::Label` in the same way (just need to convert `mykey` to a vector of bytes).
